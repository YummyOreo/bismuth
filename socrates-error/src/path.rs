#![allow(unused)]
use std::{io::stdout, path::PathBuf};

use crossterm::{
    cursor,
    event::{read, Event, KeyCode},
    execute,
    style::{Color, Print, ResetColor, SetForegroundColor},
    terminal, Result,
};

use crate::error_ui;

fn update_input(path: &mut String) -> Option<()> {
    if let Event::Key(k) = read().expect("Should be able to read") {
        match k.code {
            KeyCode::Char(x) => {
                path.push(x);
                execute!(stdout(), Print(x.to_string()));
            }
            KeyCode::Backspace => {
                path.pop();
                let print_path = format!("Enter new path: {}", path);
                execute!(
                    stdout(),
                    terminal::Clear(terminal::ClearType::CurrentLine),
                    cursor::MoveToColumn(0),
                    Print(print_path),
                );
            }
            KeyCode::Enter => return Some(()),
            _ => {}
        }
    }
    None
}

pub fn md_file_error(_description: &str) -> Option<PathBuf> {
    let options = vec![
        // "Make folder at path.".to_string(),
        "Chose another path".to_string(),
    ];

    let option = error_ui(&options)?;

    match option {
        // 1 => {}
        1 => {
            execute!(
                stdout(),
                cursor::MoveUp(u16::try_from(options.len()).unwrap() + 1_u16),
                terminal::Clear(terminal::ClearType::CurrentLine),
                cursor::SavePosition
            );
            for i in 0..options.len() + 1 {
                execute!(
                    stdout(),
                    cursor::MoveDown(1),
                    terminal::Clear(terminal::ClearType::CurrentLine),
                    cursor::RestorePosition,
                );
            }

            execute!(
                stdout(),
                ResetColor,
                Print("Enter new path: "),
                cursor::Show
            );
            let mut path = "".to_string();
            loop {
                if update_input(&mut path).is_some() {
                    break;
                }
            }
            return Some(PathBuf::from(path));
        }
        _ => panic!("Something has gone terribly wrong"),
    }
    None
}
