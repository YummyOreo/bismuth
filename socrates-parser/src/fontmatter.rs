use serde::Deserialize;
use serde_yaml::{from_str, Error};
use std::collections::BTreeMap;
use std::path::Path;

#[derive(Default, Deserialize, Debug)]
pub struct FontMatter {
    title: Option<String>,
    path: Option<String>,

    kind: Option<String>,

    values: Option<Vec<BTreeMap<String, String>>>,
}

impl FontMatter {
    pub fn new(path: &Path) -> Self {
        let title = path
            .file_name()
            .expect("Should be a file")
            .to_string_lossy()
            .to_string();

        FontMatter {
            title: Some(title),
            path: Some(path.to_string_lossy().to_string()),
            ..Default::default()
        }
    }

    pub fn update_from_str(&mut self, s: &str) -> Result<(), Error> {
        let updated_fm: FontMatter = from_str(s)?;

        if let Some(p) = updated_fm.path {
            if self.path.as_ref().unwrap_or(&"".to_string()) != &p {
                self.path = Some(p)
            }
        }

        if let Some(t) = updated_fm.title {
            if self.title.as_ref().unwrap_or(&"".to_string()) != &t {
                self.title = Some(t)
            }
        }

        let kind = updated_fm.kind.unwrap_or("default".to_string());
        if &kind != self.kind.as_ref().unwrap_or(&"".to_string()) {
            self.kind = Some(kind);
        }

        if self.values != updated_fm.values {
            self.values = updated_fm.values;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests_utils {
    use super::*;

    macro_rules! snapshot {
        ($string:tt) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../testdata/output/utils/");
            settings.bind(|| {
                insta::assert_snapshot!(format!("{:#?}", $string));
            });
        };
    }

    #[test]
    fn test_load() {
        let mut fm = FontMatter::default();
        let s_fm = "title: Test\npath: /test\nkind: test\nvalues:\n    - test: te";
        fm.update_from_str(s_fm).unwrap();

        snapshot!(fm);
    }

    #[test]
    fn test_load_1() {
        let mut fm = FontMatter {
            path: Some("/path/test".to_string()),
            ..Default::default()
        };
        let s_fm = "title: this is a title\nkind: This is another test";
        fm.update_from_str(s_fm).unwrap();

        snapshot!(fm);
    }

    #[test]
    fn test_load_2() {
        let mut fm = FontMatter {
            path: Some("/path/test".to_string()),
            ..Default::default()
        };
        let s_fm = "title: this is a title\nvalues:\n    - value-1: test\n    - value-2: test 2";
        fm.update_from_str(s_fm).unwrap();

        snapshot!(fm);
    }
}
