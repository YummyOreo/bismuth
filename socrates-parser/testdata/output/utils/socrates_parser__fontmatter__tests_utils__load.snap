---
source: socrates-parser/src/fontmatter.rs
expression: "format!(\"{:#?}\", fm)"
---
FontMatter {
    title: Some(
        "Test",
    ),
    path: Some(
        "/test",
    ),
    kind: Some(
        "test",
    ),
    values: Some(
        [
            {
                "test": "te",
            },
        ],
    ),
}
